### ISSUE TITLE
Deprecation - move to new TTS APIs

### ISSUE THREAD

**Issue by mikehardy** [ankidroid/Anki-Android#5021]
Text-to-speech APIs have moved in a way that isn't a mechanical fix - we need to look at the difference between the APIs we're using now and the new ones and migrate

**Comment by milyukhin** [comment]
Hey @mikehardy. I have a question about currently supported api versions.
In order to move to the new TTS API, first I replaced the deprecated `speak` method with the newer one and I got `Call requires API level 21 (current min is 15): android.speech.tts.TextToSpeech#speak` (the Android Studio lint warning).
So I added a if-else condition figuring which version it's running and call the according method (ended up having both methods - deprecated and the new one).
This worked fine on the api version 28, but when I was trying to run it against api 19 (on emulator) it failed with the error below.
Quistion: is api above 21 actually supported by the app?

```java
07-12 16:49:49.434 3203-3203/com.ichi2.anki E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.ichi2.anki, PID: 3203
    java.lang.ExceptionInInitializerError
        at okhttp3.OkHttpClient.<init>(OkHttpClient.kt:211)
        at okhttp3.OkHttpClient$Builder.build(OkHttpClient.kt:944)
        at com.brsanthu.googleanalytics.httpclient.OkHttpClientImpl.createHttpClient(OkHttpClientImpl.java:78)
        at com.brsanthu.googleanalytics.httpclient.OkHttpClientImpl.<init>(OkHttpClientImpl.java:36)
        at com.ichi2.anki.analytics.UsageAnalytics.initialize(UsageAnalytics.java:65)
        at com.ichi2.anki.AnkiDroidApp.onCreate(AnkiDroidApp.java:210)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1007)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4344)
        at android.app.ActivityThread.access$1500(ActivityThread.java:135)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1256)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:136)
        at android.app.ActivityThread.main(ActivityThread.java:5017)
        at java.lang.reflect.Method.invokeNative(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:515)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595)
        at dalvik.system.NativeStart.main(Native Method)
     Caused by: java.lang.IllegalStateException: Expected Android API level 21+ but was 19
        at okhttp3.internal.platform.AndroidPlatform$Companion.buildIfSupported(AndroidPlatform.kt:370)
        at okhttp3.internal.platform.Platform$Companion.findPlatform(Platform.kt:204)
        at okhttp3.internal.platform.Platform$Companion.access$findPlatform(Platform.kt:178)
        at okhttp3.internal.platform.Platform.<clinit>(Platform.kt:179)
        at okhttp3.OkHttpClient.<init>(OkHttpClient.kt:211) 
        at okhttp3.OkHttpClient$Builder.build(OkHttpClient.kt:944) 
        at com.brsanthu.googleanalytics.httpclient.OkHttpClientImpl.createHttpClient(OkHttpClientImpl.java:78) 
        at com.brsanthu.googleanalytics.httpclient.OkHttpClientImpl.<init>(OkHttpClientImpl.java:36) 
        at com.ichi2.anki.analytics.UsageAnalytics.initialize(UsageAnalytics.java:65) 
        at com.ichi2.anki.AnkiDroidApp.onCreate(AnkiDroidApp.java:210) 
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1007) 
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4344) 
        at android.app.ActivityThread.access$1500(ActivityThread.java:135) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1256) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:136) 
        at android.app.ActivityThread.main(ActivityThread.java:5017) 
        at java.lang.reflect.Method.invokeNative(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:515) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595) 
        at dalvik.system.NativeStart.main(Native Method)
```

**Comment by mikehardy** [comment]
Hey there! This one and the clipboard deprecations are probably two of the biggest API changes to handle, thanks for taking a look at it.

If TTS didn't work for all the APIs we would hear about it immediately - those users are our power users typically and raise issues when TTS breaks.

I'm guessing this is some issue with how the API-testing worked in your changes? But it's a weak guess without seeing code. Nothing wrong with posting up a 'WIP' PR that's known-broken if you're looking for feedback, if that's useful. Although I'm traveling now so may not be able to respond much in the next week or two

Maybe you could try what I refer to as "the Compat infrastructure" to implement the change to make it more obvious

That path would be to add a generic API to the Compat interface and implement the old it with the old TTS API in the V15 object, then implement it with the new TTS API in the VNN object where NN is whatever API the new API showed up. Sometimes we don't have a CompatVNN object for a given API yet because it didn't have any changes but you can follow the existing pattern to create one and add it to the infrastructure

**Comment by milyukhin** [comment]
Thanks @mikehardy 
> If TTS didn't work for all the APIs we would hear about it immediately - those users are our power users typically and raise issues when TTS breaks.

Well, in my case it did start for api 19 only... But I got what you mean.

While I'm digesting your idea of the "Compat infrastructure" (is there a place I can take a glance how it's done as an example?) I create a WIP PR for you to pre-review and general suggestions.
Btw, I not sure why you call this ticket (TTS API move) big (although, I agree that the clipboard API changes are big), as from what I found used in `ReadText.java`, deprecated was only `TextToSpeech#speak` call (can you please add on this as well)

**Comment by mikehardy** [comment]
I think of it is larger as a chunk of work only because of the testing involved - most people don't have TTS enabled and have no idea how to test it. With AnkiDroid ensuring correctness is about 90% of the work I find.

As for an example of the Compat stuff, here's one where I implemented a centralized file copy in the code base, and there was an easy API for it in API26 and a less-good one until then. So you can see work in Compat.java, CompatV15.java then CompatV26.java to implement a backwards-compatible copy, followed by it's use in the codebase https://github.com/ankidroid/Anki-Android/pull/5093/files

Same here for notification channels which was the first API26 item, so I actually created the CompatV26.java object in this PR - it's pretty clear if you focus just on the changes in the compat directory https://github.com/ankidroid/Anki-Android/pull/4922/files

**Comment by milyukhin** [comment]
Thanks Mike, I did implemented it using the Compat interface. I think it's more elegant that way.

Also I figured that bump to the latest okhttp breaks API 19 compatibility (gives the error I posted above), as per [https://github.com/square/okhttp/issues/4481](okhttp#4481) they dropped support of Android 4.
I had to downgrade those two (as they depend on okhttp 4):

> implementation 'net.mikehardy:google-analytics-java7:2.0.7'
> implementation 'com.squareup.okhttp3:okhttp:3.12.2'
